File: v2/src/main/java/com/api/v2/people/domain/Person.java
package com.api.v2.people.domain;

import com.api.v2.people.enums.Gender;
import com.api.v2.people.requests.PersonRegistrationDto;
import lombok.Getter;
import lombok.NoArgsConstructor;
import org.springframework.data.annotation.Id;
import org.springframework.data.mongodb.core.mapping.Document;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.UUID;

@Document
@Getter
@NoArgsConstructor
public class Person {

    @Id
    private String id;
    private String firstName;
    private String middleName;
    private String lastName;
    private String ssn;
    private LocalDate birthDate;
    private String phoneNumber;
    private Gender gender;
    private String email;
    private LocalDateTime createdAt;

    private Person(PersonRegistrationDto registrationDto) {
        this.id = UUID.randomUUID().toString();
        this.firstName = registrationDto.firstName();
        this.middleName = registrationDto.middleName();
        this.lastName = registrationDto.lastName();
        this.ssn = registrationDto.ssn();
        this.birthDate = registrationDto.birthDate();
        this.phoneNumber = registrationDto.phoneNumber();
        this.gender = registrationDto.gender();
        this.email = registrationDto.email();
        this.createdAt = LocalDateTime.now();
    }

    public static Person of(PersonRegistrationDto registrationDto) {
        return new Person(registrationDto);
    }

}

File: v2/src/main/kotlin/com/api/v2/people/domain/Person.kt
package com.api.v2.people.domain

import com.api.v2.people.dtos.Address
import com.api.v2.people.enums.Gender
import com.api.v2.people.requests.PersonRegistrationDTO
import org.springframework.data.mongodb.core.mapping.Document
import java.time.LocalDate
import java.time.LocalDateTime
import java.util.UUID

@Document
class Person private constructor(
    var firstName: String,
    var middleName: String?,
    var lastName: String,
    var birthDate: LocalDate,
    var sin: String,
    var email: String,
    var phoneNumber: String,
    var gender: Gender,
    var address: Address
){

    var id = UUID.randomUUID().toString();
    val createdAt: LocalDateTime = LocalDateTime.now()
    val updatedAt: LocalDateTime? = null

    companion object {
        fun of(registrationDto: PersonRegistrationDTO): Person {
            return Person(
                registrationDto.firstName,
                registrationDto.middleName,
                registrationDto.lastName,
                registrationDto.birthDate,
                registrationDto.sin,
                registrationDto.email,
                registrationDto.phoneNumber,
                registrationDto.gender,
                registrationDto.address
            )
        }
    }
}